### PR TITLE
Disk size scheduler

### PR DESCRIPTION
**Pull request by jimengliu**
1. add cache manager to load host disk information on demand. For now now resync from changes.
2. add disk size constraint and its provider to use disksize label to verify against
3. add post remove of host listener
4. add post purge of instance listener

@prachidamle @alena1108

### CODE REVIEW

**Comment by alena1108** [comment]
@jimengliu having tests would help to understand the flow; could you add them?

File: code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/DiskSizeConstraintProvider.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        @SuppressWarnings("unchecked")
+        Map<String, String> labels = DataAccessor.fields(instance).withKey(InstanceConstants.FIELD_LABELS)
+                .as(Map.class);
+        if (labels == null) {
+            return;
+        }
+
+        for (Map.Entry<String, String> labelEntry : labels.entrySet()) {
+            String labelKey = labelEntry.getKey();
+            if (labelKey.startsWith("io.rancher.scheduler.disksize.")) {
```
Comment (alena1108) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/DiskSizeConstraintProvider.java]: we should define the label prefix in SystemLabels class, and reference it here

Comment (jimengliu) [inline code/iaas/allocator/src/main/java/io/cattle/platform/allocator/constraint/DiskSizeConstraintProvider.java]: thanks, I was wondering where to put those constant strings.

**Comment by jimengliu** [comment]
added tests: 
1. one container with a volume scheduling on a host with not enough disk space will fail
2. one container with a volume scheduling on a host wit enough disk space will succeed
3. one container with 2 volumes scheduling on a host that can only satisfy one volume but not the other will fail
4. one container with 2 volumes scheduling on a host that can satisfy one volume on one of the disks, the other on another disk with enough space, will succeed

**Comment by alena1108** [comment]
@jimengliu there was something wrong with the tests, looks like it got stuck. When you run your tests locally, can you make sure that "-Didempotent.checks=true" is set as a jvm option (run->debug configuration->pick cattle app ->arguments). Thats how Drone runs them. See if its your test that gets stuck

**Comment by jimengliu** [comment]
ok, let me try that locally to see if it is stuck.

**Comment by jimengliu** [comment]
had another one sent out

**Pull request closed without merging**